proposals/cri-o.adoc

```diff
@@ -0,0 +1,221 @@
+  - github.com/modern-go/concurrent
+  - github.com/modern-go/reflect2
+  - github.com/mrunalp/fileutils
+  - github.com/opencontainers/go-digest
+  - github.com/opencontainers/image-spec
+  - github.com/opencontainers/runc
+  - github.com/opencontainers/runtime-spec
+  - github.com/opencontainers/runtime-tools
+  - github.com/opencontainers/selinux
+  - github.com/pquerna/ffjson
+  - github.com/prometheus/client_golang
+  - github.com/prometheus/client_model
+  - github.com/prometheus/common
+  - github.com/prometheus/procfs
+  - github.com/soheilhy/cmux
+  - github.com/vishvananda/netlink
+  - github.com/vishvananda/netns
+  - github.com/xeipuuv/gojsonpointer
+  - github.com/xeipuuv/gojsonreference
+  - github.com/xeipuuv/gojsonschema
+  - google.golang.org/genproto
+  - google.golang.org/grpc
+  - gopkg.in/square/go-jose.v2
+  - gopkg.in/yaml.v2
+  - k8s.io/api
+  - k8s.io/apiextensions-apiserver
+  - k8s.io/apimachinery
+  - k8s.io/apiserver
+  - k8s.io/client-go
+  - k8s.io/csi-api
+  - k8s.io/kube-openapi
+  - k8s.io/kubernetes
+  - k8s.io/utils
+* BSD-2
+  - github.com/godbus/dbus
+  - github.com/pkg/errors
+  - github.com/seccomp/libseccomp-golang
+  - github.com/syndtr/gocapability
+  - gopkg.in/mgo.v2
+* BSD-3
+  - github.com/PuerkitoBio/purell
+  - github.com/PuerkitoBio/urlesc
+  - github.com/cyphar/filepath-securejoin
+  - github.com/fsnotify/fsnotify
+  - github.com/gogo/protobuf
+  - github.com/golang/protobuf
+  - github.com/gorilla/context
+  - github.com/gorilla/mux
+  - github.com/imdario/mergo
+  - github.com/mtrmac/gpgme
+  - github.com/pmezard/go-difflib
+  - github.com/spf13/pflag
+  - github.com/ulikunitz/xz
+  - github.com/vbatts/tar-split
+  - golang.org/x/crypto
+  - golang.org/x/net
+  - golang.org/x/oauth2
+  - golang.org/x/sys
+  - golang.org/x/text
+  - golang.org/x/time
+  - gopkg.in/cheggaaa/pb.v1
+  - gopkg.in/fsnotify.v1
+  - gopkg.in/inf.v0
+  - gopkg.in/tomb.v1
+* MIT
+  - github.com/Azure/go-ansiterm
+  - github.com/BurntSushi/toml
+  - github.com/Microsoft/go-winio
+  - github.com/Microsoft/hcsshim
+  - github.com/asaskevich/govalidator
+  - github.com/beorn7/perks
+  - github.com/blang/semver
+  - github.com/buger/goterm
+  - github.com/dgrijalva/jwt-go
+  - github.com/docker/docker-credential-helpers
+  - github.com/emicklei/go-restful
+  - github.com/emicklei/go-restful-swagger12
+  - github.com/exponent-io/jsonpath
+  - github.com/fatih/camelcase
+  - github.com/ghodss/yaml
+  - github.com/go-zoo/bone
+  - github.com/gregjones/httpcache
+  - github.com/hpcloud/tail
+  - github.com/json-iterator/go
+  - github.com/kr/pty
+  - github.com/mailru/easyjson
+  - github.com/mattn/go-runewidth
+  - github.com/mitchellh/go-wordwrap
+  - github.com/mitchellh/mapstructure
+  - github.com/peterbourgon/diskv
+  - github.com/renstrom/dedent
+  - github.com/sirupsen/logrus
+  - github.com/tchap/go-patricia
+  - github.com/ugorji/go
+  - github.com/urfave/cli
+* LGPLv3
+  - github.com/juju/ratelimit
```

Comment:
This is a problem -- in Go libraries are not replaceable and thus LGPLv3 requires that the entire project be LGPLv3 licensed. The exceptions to allow multiple licenses only work if you can replace them.

This should be a worry for everyone distributing cri-o since currently cri-o is actually LGPLv3 because of this dependency...

Reply:
IANAL and will seek guidance this. Though as I read it, the supplemental permisson of this LGPLv3 allowed works that convert or combine with this library source code, and do not impose the demand of section 3 and 4 of the GPLv3, as long as that code is unmodified, license and access to it's source code is provided. Thereby not making cri-o an LGPv3 work.

Reply:
@vbatts explanation is how I see things

However, can cri-o replace it's usage of juju/ratelimit with golang.org/x/time/rate
https://github.com/kubernetes-sigs/cri-o/issues/1935 https://github.com/kubernetes/kubernetes/pull/38320

Reply:
Reading the LGPLv3 again, &sect; 4.d.0 allows you to distributed combined works with an LGPLv3 project so long as you provide the application source code under a license that allows combination with the library. I was basing my comment on LGPLv2.1 but obviously it handles this slightly differently.

I'm not a lawyer either, but I didn't see wording to the effect of what @vbatts said -- but at the end of the day &sect 4.d.0 is sufficient.

Reply:
Yeah, we will switch to the other library like k8s did.